community/neo4j/src/test/java/ConcurrentChangesOnEntitiesTest.java

```diff
@@ -0,0 +1,240 @@
+import org.neo4j.graphdb.DynamicLabel;
+import org.neo4j.graphdb.DynamicRelationshipType;
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Node;
+import org.neo4j.graphdb.Relationship;
+import org.neo4j.graphdb.Transaction;
+import org.neo4j.graphdb.factory.GraphDatabaseFactory;
+import org.neo4j.helpers.progress.ProgressMonitorFactory;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.logging.FormattedLogProvider;
+import org.neo4j.logging.LogProvider;
+import org.neo4j.test.TargetDirectory;
+
+import static org.junit.Assert.assertTrue;
+
+public class ConcurrentChangesOnEntitiesTest
+{
+    @Rule
+    public TargetDirectory.TestDirectory testDirectory = TargetDirectory.testDirForTest( getClass() );
+
+    @Test
+    public void addConcurrentlySameLabelToANode() throws Throwable
+    {
+        final GraphDatabaseService db = new GraphDatabaseFactory().newEmbeddedDatabase( testDirectory.graphDbDir() );
+
+        final long nodeId = initWithNode( db );
+        final AtomicReference<Exception> ex = new AtomicReference<>();
+
+        Thread t1 = new Thread( new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try ( Transaction tx = db.beginTx() )
+                {
+                    Node n = db.getNodeById( nodeId );
+                    n.addLabel( DynamicLabel.label( "A" ) );
+                    tx.success();
+                }
+                catch ( Exception e )
+                {
+                    ex.set( e );
+                }
+            }
+        } );
+
+        Thread t2 = new Thread( new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try ( Transaction tx = db.beginTx() )
+                {
+                    Node n = db.getNodeById( nodeId );
+                    n.addLabel( DynamicLabel.label( "A" ) );
+                    tx.success();
+                }
+                catch ( Exception e )
+                {
+                    ex.set( e );
+                }
+            }
+        } );
+
+        t1.start();
+        t2.start();
+
+        t1.join();
+        t2.join();
+
+        if ( ex.get() != null )
+        {
+            throw ex.get();
+        }
+
+        db.shutdown();
+
+        LogProvider logProvider = FormattedLogProvider.toOutputStream( System.out );
+        try
+        {
+            ConsistencyCheckService.Result result = new ConsistencyCheckService().runFullConsistencyCheck(
+                    testDirectory.graphDbDir(), new Config(), ProgressMonitorFactory.textual( System.err ),
+                    logProvider );
+            assertTrue( result.isSuccessful() );
+        }
+        catch ( ConsistencyCheckIncompleteException e )
+        {
+            throw new RuntimeException( e );
+        }
+    }
+
+    @Test
+    public void setConcurrentlySamePropertyWithDifferentValuesOnANode() throws Throwable
```

Comment:
Since there is nothing coordinating the two threads, there is no guarantee that the issue will manifest.